src/System.Private.Interpreter/src/Internal/Runtime/Interpreter/ILInterpreter.cs

```diff
@@ -2126,5 +2116,370 @@ private void InterpretSwitch(ref ILReader reader, int jmpBase, int[] jmpDelta)
                     reader.Seek(jmpBase + jmpDelta[i]);
             }
         }
+
+        private void InterpretNewArray(int token)
+        {
+            int length = 0;
+            StackItem stackItem = PopWithValidation();
+
+            switch (stackItem.Kind)
+            {
+                case StackValueKind.Int32:
+                    length = stackItem.AsInt32();
+                    break;
+                case StackValueKind.NativeInt:
+                    length = (int)stackItem.AsNativeInt();
+                    break;
+                default:
+                    ThrowHelper.ThrowInvalidProgramException();
+                    break;
+            }
+
+            Debug.Assert(length >= 0);
+
+            TypeDesc arrayType = ((TypeDesc)_methodIL.GetObject(token)).MakeArrayType();
+            Array array = RuntimeAugments.NewArray(arrayType.GetRuntimeTypeHandle(), length);
```

Comment:
This doesn't work for non primitive types. Something like `new DateTime[10]` fails because the runtime is unable to find the `RuntimeTypeHandle` for `DateTime[]` even though there exists one for `DateTime`. Is there a way to probe the metadata of the interpreted assembly (and its dependencies) for runtime type information?

Reply:
It should work for arbitrary reference type arrays - value types are harder.

You should be able to work around by adding this array to RD.XML or adding `typeof(DateTime[])` somewhere in the code that hosts the interpreter.